### PR TITLE
Release 1.4.2: UHID HID input, overlay cursor, and pointer compensation

### LINKED ISSUE #18: Add configurable virtual pointer sensitivity
Hi @anasvhora284, this follow-up comes from my review of relative mouse handling in #10.

## Background

Input Leaf currently forwards relative X/Y movement to the virtual UHID mouse and clamps each report to the supported HID range. A virtual mouse has no physical DPI, but the app could offer an equivalent pointer-sensitivity multiplier.

## Goal

Let users adjust virtual pointer sensitivity while keeping the current unscaled behavior as the default.

## Proposed scope

- [ ] Define the sensitivity range, default value, and user-facing terminology.
- [ ] Decide where scaling belongs so behavior remains consistent across supported input methods.
- [ ] Add preference persistence and settings UI.
- [ ] Scale relative X/Y movement without losing small fractional movement.
- [ ] Split or accumulate scaled deltas that exceed a single HID report instead of silently losing movement through clamping.
- [ ] Preserve existing behavior when sensitivity is set to the default.
- [ ] Add deterministic tests for minimum, default, maximum, fractional, negative, and overflow cases.
- [ ] Document interaction with operating-system pointer acceleration.

## Non-goals

- Claiming that the virtual device exposes physical hardware DPI.
- Requiring a physical mouse in the normal pull-request test suite.
- Adding trackpad or multitouch gestures; those are tracked separately in #17.

The implementation should use terms such as “pointer sensitivity” rather than physical DPI unless the underlying virtual device eventually exposes a real DPI control.

### LINKED ISSUE #20: Add framing and versioning to the app-to-UHID event protocol
Hi @anasvhora284, this follow-up comes from strengthening UHID event decoding in #10.

## Background

The current local-socket protocol sends an event type followed by a fixed-size payload. It has no frame length or protocol version. The decoder can identify a truncated event, but an extra or missing byte may misalign the stream and cause every following byte to be interpreted as the wrong event.

The socket is local and peer-verified, so this is mainly a reliability and compatibility concern rather than a request for network-level complexity.

## Goal

Define a small framed and versioned event format so malformed or future events can be rejected or skipped without corrupting the rest of the stream.

## Proposed scope

- [ ] Inventory the app-side writer and UHID-side reader for every current event.
- [ ] Define a compact header containing at least protocol version, event type, and payload length.
- [ ] Set a conservative maximum payload length before allocating or reading data.
- [ ] Validate exact payload lengths for known event types.
- [ ] Decide whether unknown event types can be skipped safely by length.
- [ ] Include shutdown behavior in the framed protocol design.
- [ ] Define compatibility and rollout behavior while the app and generated UHID DEX may be on different versions.
- [ ] Add deterministic codec tests for valid, truncated, oversized, unknown, and consecutive frames.
- [ ] Test that one malformed frame does not misalign the next valid frame where recovery is supported.
- [ ] Update the generated UHID DEX asset through the normal build task.

## Non-goals

- Adding encryption or checksums without a demonstrated need on the peer-verified local socket.
- Depending on Android sockets or hardware in ordinary JVM tests.
- Mixing pointer sensitivity or scroll-speed settings into the wire-format change.

The format should remain intentionally small and should be agreed before implementation because both sides must migrate together.

### LINKED ISSUE #28: IME layouts (Gujarati Inscript, etc.) send no KeyDown from Deskflow
XKB layouts (Russian, Gujarati Phonetic, etc.) work: Deskflow sends a KeyID and we inject Unicode or a physical shortcut.

IME layouts such as GNOME IBus **Gujarati (Inscript)** do not. The host IME consumes the key before Deskflow/libei. Input Leap skips `keyid == 0` (`NoSymbol`), so the Android client never receives a `KeyDown`. Logcat from a 2026-08-26 device test showed ~35s on the phone with **zero** letter events, then Phonetic `0x61` (`a`) and Russian `0x444` (`ф`) plus Ctrl+A working.

This is not a client keymap gap. Input Leaf cannot invent keys the server never sent.

**Workaround:** use an XKB source (e.g. Indian → Gujarati / `in(guj)`), not IBus `gu`. Restart Deskflow after switching.

**Possible later work (host/Deskflow):** forward physical `KeyButton` even when the keysym is `NoSymbol`, or capture keys before the IME. Same class of failure for other IBus/fcitx Indic (and similar) layouts.

Tracked out of #5 / #22 so that PR can merge.

### LINKED ISSUE #34: 无法使用电脑键盘在手机上输入中文
无法使用电脑键盘在手机上输入中文

### PR DESCRIPTION
**Pull request by anasvhora284**
## Overview
Release 1.4.2 represents a major architectural and functional upgrade for Input Leaf. It introduces native kernel-level UHID hardware keyboard and mouse emulation via Shizuku, dynamic Android system pointer speed compensation, mid-session display rotation handling, rich in-app Markdown release notes, and accessibility overlay reliability improvements, while completely removing legacy daemon sidecars.

---

## Features Added

- **Native Hardware Keyboard & Mouse Emulation (UHID via Shizuku)**
  - Attaches genuine Linux kernel HID keyboard and mouse devices when the cursor enters the phone screen and cleanly detaches on Leave.
  - Enables Android's native physical hardware keyboard subsystem, letting Gboard (and other system IMEs) display the on-screen candidate composition bar for **Chinese Pinyin, Gujarati Inscript, Indic, Russian**, and other language layouts.

- **System Pointer Speed Synchronization & Sub-Pixel Tracking**
  - Directly observes Android's native mouse pointer speed setting (`Settings.System.POINTER_SPEED`, range `-7` to `+7`).
  - Implements dynamic gain curve compensation and accumulates fractional sub-pixel deltas so the pointer responds accurately to the user's system mouse preferences without needing a redundant in-app slider.

- **Mid-Session Dynamic Display Rotation (`DINF`)**
  - Sends live screen dimensions and orientation updates over the Input Leap connection on screen rotation mid-session.
  - Keeps coordinate mapping, pointer clamping, and edge anchoring aligned without requiring a disconnect/reconnect.

- **Rich Markdown Changelog in Update Dialog**
  - Integrated a lightweight, zero-dependency Compose Markdown engine (`MarkdownContent.kt`) that renders GitHub release notes with full support for headings, bullet/numbered lists, bold/italic styling, inline code, and clickable URLs.

- **Accessibility & Connection UX Enhancements**
  - Accessibility overlay cursor remains visible and reliable without treating `SYSTEM_ALERT_WINDOW` as an exclusive blocker.
  - Server cards show clear real-time connecting states; first Enter after connecting no longer suffers from latency freezes; and Shizuku binders re-attach reliably across app reboots.

---

## Technical & Architectural Improvements

- **In-Process `/dev/uhid` Communication**: Completely removed the standalone `uhid-server` Java DEX sidecar daemon and unframed local socket IPC. Input Leaf now writes standard Linux kernel `uhid_event` structs directly into `/dev/uhid` via Shizuku (`UhidChannel.kt`, `UhidProtocol.kt`).
- **Direct Evdev-to-HID Scancode Pipeline**: Replaced legacy keysym translation tables with `EvdevToHid.kt` and standard USB HID keyboard reports (`HidKeyboard.kt`), providing 1:1 hardware key injection.
- **Modular Pointer Engine**: Added `HidMouseState.kt` (delta accumulation), `MouseEdgeAnchor.kt` (edge snapping and re-entry), `MousePointerCompensation.kt` (OEM curve compensation including OnePlus/OPPO offsets), and `HidAttachmentController.kt`.
- **Codebase Pruning**: Removed obsolete setup wizards, unused resolvers, legacy patch files (`uhid_mouse_backup.patch`), and stale documentation (`DEVELOPMENT_JOURNEY.md`, `UI_REDESIGN.md`).
- **Expanded Unit Test Suite**: Added comprehensive test coverage across `UhidProtocolTest`, `HidKeyboardTest`, `HidMouseTest`, `MouseEdgeAnchorTest`, `MousePointerCompensationTest`, `PointerSpeedTest`, and `MarkdownParserTest`.

---

## Issues Closed & Resolved

- **Closes #18** (Configurable pointer sensitivity) — Virtual mouse movement dynamically hooks into the native Android system setting (`Settings.System.POINTER_SPEED`, range `-7` to `+7`) with active compensation curves and fractional sub-pixel accumulation in `HidMouseState`.
- **Closes #20** (App-to-UHID event protocol framing) — Superseded by the complete removal of the standalone `uhid-server` Java DEX sidecar daemon and local socket IPC in favor of direct `/dev/uhid` via Shizuku.
- **Closes #28** (IME layouts like Gujarati Inscript / Indic dropping keys) — Resolved via physical hardware keyboard emulation. Users keep the host keyboard on standard QWERTY and configure the desired layout in Android's physical keyboard settings (`Settings → System → Physical keyboard → Gboard`), allowing Android to compose Gujarati, Indic, and other language layouts natively.
- **Closes #34** (Chinese IME input via computer keyboard) — Resolved via physical keyboard attachment. Android recognizes the phone as having an external hardware keyboard, enabling Gboard's native physical keyboard IME pipeline (Pinyin composition bar and candidate selection) directly on-device.

*(Note: #18, #28, and #34 are resolved in Shizuku / UHID mode where a real kernel device is recognized by Android. Pure Accessibility fallback remains touch-drag and InputLeafIME-based).*

---

## Test Plan & Verification

- [x] **Overlay Cursor**: Visible and functional in accessibility mode without treating overlay permission as an exclusive gate.
- [x] **UHID Device Lifecycle**: HID keyboard and mouse attach cleanly on Enter and disconnect immediately on Leave.
- [x] **Rotation Mid-Session**: Live screen dimensions update via DINF and pointer bounds remain 1:1 after rotating phone.
- [x] **Pointer Tracking**: Tracks smoothly across DPIs with sub-pixel compensation and edge anchoring.
- [x] **Connection Stability**: Server cards reflect connecting state; first Enter does not freeze; Shizuku reconnects reliably.
- [x] **Rich Changelog**: Markdown parser formats headers, bullet lists, bold/italic, code, and links in update dialog.
- [x] **Device Verified**: Tested and verified live on Android 16 (OnePlus Nord 4).

### CODE REVIEW

File: app/src/main/java/com/inputleaf/android/shizuku/uhid/UhidProtocol.kt [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+ * The field that is easy to get wrong: `rd_size` is a **u16 at offset 260**, after
+ * `name[128]`, `phys[64]` and `uniq[64]`. Writing it as a u32 at 132 lands inside
+ * `phys[]` and the kernel rejects or misparses the device.
+ */
+internal object UhidProtocol {
+
+    const val UHID_DESTROY = 1
+    const val UHID_START = 4
+    /** EventHub opened the evdev node — safe to send the first input report. */
+    const val UHID_OPEN = 6
```
Comment (guaje) [inline app/src/main/java/com/inputleaf/android/shizuku/uhid/UhidProtocol.kt]: **issue (spec_mismatch):** The event-type constants do not match `include/uapi/linux/uhid.h`, where `UHID_START = 2`, `UHID_STOP = 3`, `UHID_OPEN = 4`, `UHID_CLOSE = 5`, `UHID_OUTPUT = 6`. The `UHID_START = 4` above is really the kernel's **OPEN**, and the `UHID_OPEN = 6` is really the kernel's **UHID_OUTPUT** (a rare host→device event). The readiness wait therefore blocks on the correct practical signal only by coincidence, the `sawOpen` fast-path in `UhidChannel` is dead against real kernels (readiness always degrades to the timeout + presence-probe path), and `payloadSize()` classifies the real `UHID_OUTPUT(6)` as zero-payload, which misframes the read stream for any code path that drains events by type. The unit tests inject these same constants back into fake streams, so CI validates internal consistency rather than kernel correctness.

**Triggers:** Any real device: kernel `START(2)` is ignored, kernel `OPEN(4)` is consumed as “START”, the “OPEN” branch never fires; a maintainer following the doc comments reasons incorrectly about the ABI.

**Suggested fix:** Align the constants with `uhid.h`, decide explicitly whether readiness waits on `START(2)` or `OPEN(4)` and name the constant accordingly, correct the `payloadSize()` table (`OUTPUT` carries `u16 size + data[4096]`), and re-base the readiness tests on the kernel values so they guard the ABI.

File: app/src/main/java/com/inputleaf/android/ui/components/MarkdownContent.kt [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+
+            val token = match.value
+            when {
+                // Markdown link: [text](url)
+                token.startsWith("[") && token.contains("](") && token.endsWith(")") -> {
+                    val label = token.substringAfter("[").substringBefore("](")
+                    val url = token.substringAfter("](").substringBeforeLast(")")
+                    withLink(
+                        LinkAnnotation.Url(
+                            url = url,
```
Comment (guaje) [inline app/src/main/java/com/inputleaf/android/ui/components/MarkdownContent.kt]: **issue (security):** The changelog body is fetched from the GitHub release API, and this branch passes the Markdown-link URL into `LinkAnnotation.Url` with no scheme validation, while the raw-URL branch right below restricts to `http(s)`. A tampered, typosquatted, or compromised release body can emit `intent://` or other-scheme links; the default `AndroidUriHandler` dispatches them via `ACTION_VIEW`, so one tap in the update dialog becomes an arbitrary implicit intent.

**Triggers:** Release notes contain a `[label](non-http-scheme-url)` link and the user taps it in the update dialog.

**Suggested fix:** Apply the same http(s) check before constructing the link, falling back to plain text:

```kotlin
val url = token.substringAfter("](").substringBeforeLast(")")
if (url.startsWith("http://") || url.startsWith("https://")) {
    withLink(LinkAnnotation.Url(url = url, styles = …)) { append(label) }
} else {
    append(label)
}
```

File: app/src/main/java/com/inputleaf/android/service/ConnectionService.kt [hunk 3]
```diff
@@ -422,7 +585,8 @@ class ConnectionService : Service() {
     private fun scheduleRetry(ip: String, screenName: String, generation: Int) {
         if (userInitiatedDisconnect || generation != connectGeneration) return
         retryJob?.cancel()
-        val delayMs = RetryDelayCalculator.getDelay(retryAttempt++)
+        val delayMs = RETRY_DELAYS_MS[retryAttempt.coerceIn(0, RETRY_DELAYS_MS.lastIndex)]
```
Comment (guaje) [inline app/src/main/java/com/inputleaf/android/service/ConnectionService.kt]: **issue (broader_impact):** The retry-delay selection was inlined here when `RetryDelayCalculator` was deleted, and `RetryDelayCalculatorTest` (boundary attempts 0/1/2/3/4/5/10 and negative input) went with it. This line lives inside a `Service` method, which no JVM test can reach, so the exact-delay-per-attempt behavior is no longer asserted anywhere: the 100% patch gate passes because the moved code is no longer measured, not because it is covered. This also reverses the direction this codebase has been moving (small testable collaborators extracted out of the Service, e.g. #42's `ShizukuActiveSessionPolicy`).

**Triggers:** Any future edit to the `RETRY_DELAYS_MS` table or the `coerceIn` bounds can regress silently through `fast-jvm`.

**Suggested fix:** Keep the delay table as a private val if desired, but restore the lookup as a standalone `internal object` (or internal function) with its boundary tests back in the JVM suite.

File: app/src/main/java/com/inputleaf/android/shizuku/ShizukuInputInjector.kt [hunk 4]
```diff
@@ -127,114 +162,323 @@ class ShizukuInputInjector(
...
-    /**
-     * Unbind from the Shizuku service.
-     */
+
     override fun disconnect() {
         if (isBound || service != null) {
             try {
+                runCatching { service?.releaseHidKeys() }
+                runCatching { service?.closeVirtualKeyboard() }
+                runCatching { service?.closeVirtualMouse() }
```
Comment (guaje) [inline app/src/main/java/com/inputleaf/android/shizuku/ShizukuInputInjector.kt]: **issue (broader_impact):** When the Shizuku binder dies (`ConnectionService.handleShizukuDied()` → `disconnect()`), the explicit UHID teardown cannot reach the UserService process — the `closeVirtualKeyboard()`/`closeVirtualMouse()` binder calls throw `DeadObjectException` and the `runCatching` wrappers swallow it — so no `UHID_DESTROY` is ever written. Kernel cleanup then depends entirely on `unbindUserService(force)` reaping the process promptly; on OEM Shizuku variants with slow teardown the HID keyboard/mouse stay attached to Android, which believes a physical keyboard is present and keeps the soft keyboard suppressed.

**Triggers:** The Shizuku server process dies (or is killed) while UHID devices are attached.

**Suggested fix:** Have the UserService `linkToDeath` on the client binder (or on its own `IService` token) and close its `UhidChannel`s directly in the death callback, so the kernel destroys the devices without depending on process-kill timing.

File: app/src/main/java/com/inputleaf/android/shizuku/ShizukuInputInjector.kt [hunk 5]
```diff
@@ -127,114 +162,323 @@ class ShizukuInputInjector(
...
+                    hidMouse.setButtons(buttons)
+                    svc.injectHidMouse(0, 0, buttons, 0)
+                }
+                is InputLeapEvent.MouseUp -> {
+                    val buttons = hidMouse.buttons() and inputLeapButtonToHid(event.buttonId).inv()
+                    hidMouse.setButtons(buttons)
+                    svc.injectHidMouse(0, 0, buttons, 0)
+                }
+                is InputLeapEvent.MouseWheel ->
+                    svc.injectHidMouse(0, 0, hidMouse.buttons(), event.yDelta / 120)
```
Comment (guaje) [inline app/src/main/java/com/inputleaf/android/shizuku/ShizukuInputInjector.kt]: **nitpick:** Integer division truncates wheel notches: any `yDelta` that is not a multiple of 120 becomes `0` on the HID path (e.g. `60 → 0`), while the accessibility branch below uses float division. The Input Leap protocol normalizes wheel deltas to 120-units so real-world exposure is limited, but non-conforming servers lose scroll input silently.

Three more low-severity notes in the same area, listed here to keep them together:
1. `ConnectionService.kt` — `screenWidth`/`screenHeight` are plain vars written on the main thread (rotation/`DINF` path) and read from the IO event loop; no happens-before edge, so the fallback touch path can clamp against stale bounds (the HID path is synchronized via `HidMouseState.resizeDisplay`). Consider `@Volatile`.
2. `MousePointerCompensation.kt` — `gainForSpeedMmPerS()` has an unreachable fallback after the `+Inf`-terminated curve segment.
3. `HidAttachmentControllerTest` / `UhidChannelReadinessTest` — `Thread.sleep(20–40ms)` feeder threads race wall-clock timing (flake risk); `UhidReadinessConfig` already exposes injectable `sleeper`/`nanoTime` seams — driving determinism through them avoids the sleeps.

**Suggested fix:** Accumulate the wheel remainder (or round to nearest notch) on the HID path; apply the three notes opportunistically.

**Review by guaje — COMMENTED** [review]
Review of the 1.4.2 UHID subsystem and changelog renderer — 5 findings, anchored inline.

- `UhidProtocol.kt:20` — UHID event-type constants diverge from the kernel ABI; the readiness wait happens to work on the wrong-but-adjacent kernel signal, and the unit tests cannot catch it (issue)
- `MarkdownContent.kt:169` — Markdown-link URLs from the remote release body bypass the http(s) restriction applied to raw URLs (issue, security)
- `ConnectionService.kt:588` — retry-delay logic inlined into a Service method and its boundary tests deleted; the patch gate now passes because the logic is no longer measured (issue)
- `ShizukuInputInjector.kt:170` — phantom UHID keyboard/mouse can persist after binder death because kernel cleanup relies on process-kill timing (issue)
- `ShizukuInputInjector.kt:371` — wheel-notch truncation plus three smaller low-severity notes (nitpick)

For balance: the `UHID_CREATE2` byte layout was checked field-by-field against `uhid.h` and is exact (incl. `rd_size` u16 at 260), the module removal and CI teardown are coherent, and the coverage gates were not relaxed.

**Comment by anasvhora284** [comment]
@guaje I have made the changes according to your findings and some bugs i encountered

**Review by guaje — APPROVED** [review]
Approved — all five previous-round findings are verifiably fixed at `53308c9`, and the fix round holds up under a second multi-lens pass (kernel ABI checked against `uhid.h`/`uhid.c`, lock analysis on the new death-watch, full read of the reader lifecycle). CI green (fast-jvm, android-coverage, codecov/patch). No blockers.

**Optional to fold in before tagging 1.4.2** — one word, closes an app-crash path: in `ShizukuInputInjector.onHidMouseEnter`, the generic `catch` runs `handleRemoteException()` and then *falls through* into `finishPendingSnap(svc)` using the binder just declared dead; a `DeadObjectException` from `injectHidMouse` there escapes into the event-loop collector in `ConnectionService`, which has no catch, and reaches the uncaught handler. The `DeadObjectException` branch already `return`s — do the same after `handleRemoteException()`. If folded in, no follow-up issue is needed; otherwise I will track it as one.

**Readiness gate — evaluated, recommended as a fast follow rather than a pre-tag fold-in.** HEAD's `UhidChannel.awaitReady` waits on OPEN alone, and evaluation shows this is not confined to the previously documented OEM family:

- Kernel mechanics (`drivers/hid/uhid.c`): `UHID_START` is queued during probe (`uhid_hid_start` → inside `hid_hw_start`, before the evdev node exists) and is guaranteed whenever the device bound; `UHID_OPEN` is queued only when userspace *opens* the evdev node — stock InputReader opens every scanned device, but nothing in the kernel guarantees an opener exists, and OEM policy that filters or delays input-device opens is precisely the variance this app's device matrix contains.
- The repo's own ColorOS rationale does not actually establish an OPEN-less device: under the corrected constants, what the old code "saw as START" was the real OPEN, and the event it "never saw" was value 6 = `UHID_OUTPUT`, which never arrives on any device (it is host→device).
- Current behavior regresses attach latency on both device classes: (1) OPEN-less devices burn the full `openTimeoutMs` on every `createDevice` — even though the reader already tracks START (`UhidChannel.kt:220`, and there is even a log branch for it), because the loop condition at `:169` checks only `sawOpen`; (2) OPEN-present devices now run the sysfs presence confirmation *after* OPEN (`:173-178`), adding up to `presenceTimeoutMs`, where the pre-fix loop treated presence as an *alternative* exit so whichever signal came first won.
- Functional impact is bounded — `createDevice` proceeds either way and reports flow — which is why this is a follow-up, not a blocker. The correct predicate looks like `sawStart || sawOpen` with presence demoted to a logged confirmation; the plumbing already exists. Recommend pairing it with virtual-clock tests for both exits rather than a pre-tag change to every device's attach path.

Follow-up issues will be filed right after merge: reader-thread stop lifecycle (the `stopReader()`/`interrupt()`/dup-PFD cluster in `ec43809`), the readiness gate above, the dead stream-model paths (`waitForStart`/`drainPayload` + the unreachable deferred-INPUT2 queue), test timing patterns, and a low-severity bundle. The inline-Service findings (`ConnectAttemptOutcome`/`shouldClearActiveSession`, idle-detach logic) are deliberately not filed — they will be re-extracted by #32's rebase onto this merge, as noted on that PR.

**Comment by rlqnwkd** [comment]
Is it possible to reflect the mobile screen size in real-time—specifically the folded state—similar to Samsung's Multi Control?

**Comment by anasvhora284** [comment]
@rlqnwkd This ships with the 1.4.2 release.

When the phone switches between the cover display and the inner display, Input Leaf sends the new screen size to the server on the open connection. A half-fold on the same inner screen does not change the reported size, so that posture stays as it is.
